Updated output struct fields

mike-w-wilson opened the pull request:
Removed unnecessary annotations from output sample struct.

mike-w-wilson committed:
Updated output struct fields

lfrancioli reviewed (commented):
I can't comment on all lines because this PR only encompasses a  few changes, so I'll do it here:
- you should `filter_cols(hl.is_defined(mt.cram))` instead of filering for it in your  expr.
- for males in `get_expr_for_hemi`, you  should think whether you want to reduce the `dp_threshold` to half  since they are haploid.
- you should use `gnomad_qc.vv3.resources` instead of redifining any  path to resources. In particular:
   -`get_full_mt` instead `mt = hl.read_matrix_table(args.variant_mt_path).key_rows_by("locus", "alleles")` as it avoids passing through the data to verify sort order. 
   - the cram_paths would be good to add to the gnomad_qc v3 repo
- remove `densify`
- you compute `adj` but never use it  --remove
- use `filter_entries((mt.GQ >= gq_threshold) & (mt.DP >= dp_threshold)` (possibly with a difference for male at non-PAR for the `dp_threshold` and remove from your expressions' right now you're computing them 3x/entry
- only run the hemi logic on X/Y non-PAR. (otherwise return missing)
- Do you need to produce a single list samples with the `var_type`? It would likely be better to produce 3 separate lists of samples and forgo the addition of a string for each GT. 
- This is likely a small thing, but it you need the single list, it' s probably better to add that `var_type` after the `take` (I'd add it in the last `select` when you join the lists together)
- You forgot to add the hemi vars in the last `select`

mike-w-wilson committed:
Addressed initial feedback

mike-w-wilson commented:
Thanks for the feedback @lfrancioli !

> * you should `filter_cols(hl.is_defined(mt.cram))` instead of filering for it in your  expr.

This makes sense and I've moved it to the same filter as release-able samples

> * for males in `get_expr_for_hemi`, you  should think whether you want to reduce the `dp_threshold` to half  since they are haploid.

I did address this but I'm not thrilled with my approach, feedback welcome.

> * you should use `gnomad_qc.vv3.resources` instead of redifining any  path to resources. In particular:
>   -`get_full_mt` instead `mt = hl.read_matrix_table(args.variant_mt_path).key_rows_by("locus", "alleles")` as it avoids passing through the data to verify sort order.

This makes sense and I've adjusted the metadata table as well

>   * the cram_paths would be good to add to the gnomad_qc v3 repo

These are going to be deleted so I can add this to the repo but they will not be relevant at the end of the month.

> * remove `densify`

👍 

> * you compute `adj` but never use it  --remove

👍 

> * use `filter_entries((mt.GQ >= gq_threshold) & (mt.DP >= dp_threshold)` (possibly with a difference for male at non-PAR for the `dp_threshold` and remove from your expressions' right now you're computing them 3x/entry

Used this instead but again I dont love the dp threshold expr approach. Thanks for including why this is a bad move with the 3x explanation!


> * only run the hemi logic on X/Y non-PAR. (otherwise return missing)

This makes sense!

> * Do you need to produce a single list samples with the `var_type`? It would likely be better to produce 3 separate lists of samples and forgo the addition of a string for each GT.

This was a request of @bw2 . I believe his pipeline needs them?
> * This is likely a small thing, but it you need the single list, it' s probably better to add that `var_type` after the `take` (I'd add it in the last `select` when you join the lists together)
> * You forgot to add the hemi vars in the last `select`

I'm not sure I understand here, the hemi array is the last extend?
```
ht = ht.select(
        samples=ht.samples_w_het_var.extend(
            ht.samples_w_hom_var.extend(ht.samples_w_hemi_var)
        )
    )
```

lfrancioli commented:
@bw2 : Can you  confirm that you need a single annotation with an array with elements of type struct{s, GQ, var_type (het, hom, hemi)} and not 3 annotations het, hom, hemi each with an array with elements of type struct{s, GQ}.

bw2 commented:
@lfrancioli I wanted the single array of struct{s, GQ, var_type (het, hom, hemi)} because I was thinking the last step afterward would be to explode the table on that array, and write out the ~1.05 billion records to many tsv's (ideally in a  way where each tsv has all the records from variants with pos % N == k where N = 20,000 - so 20,000 tsv per chromosome).
I have no idea if hail can handle that currently? or if you guys can think of a better way to send this info into cromwell/Terra?

lfrancioli commented:
I see! I think Hail should be able to handle this -- that said you don't need the single array to explode; you can just combine them on the fly in an annotate before explode. If we're going to keep the Hail Table for future reference, I think it'd make more sense to leave them as 3 arrays since it saves space and potentially is easier to query.

lfrancioli commented on readviz_prep.py:
```diff
@@ -14,110 +15,98 @@
-
+    logger.info("Filtering to entries meeting GQ and DP thresholds")
+    mt = mt.filter_entries(
+        (mt.GQ >= gq_threshold) & (mt.DP >= dp_threshold_expr(mt, dp_threshold))
```
add `mt.GT.is_non_ref()` ?

lfrancioli replied:
so hom ref GTs are filtered once (vs 3x)

lfrancioli reviewed (approved):
See last small optimization and comment regarding 1 vs 3 arrays in discussion. I approved already as I think it's good but happy to take another look if you decide to implement the 3 arrays solution.

mike-w-wilson commented:
Thanks @lfrancioli! @bw2 , are you fine with the three list approach and annotating variant type and creating the single array in the next steps to avoid storing the extra string per struct entry?

mike-w-wilson commented:
Thanks @lfrancioli! @bw2 , are you fine with the three list approach and annotating variant type and creating the single array in the next steps to avoid storing the extra string per struct entry?

bw2 commented:
Storing 3 arrays and computing the single array afterward sounds good to me.

mike-w-wilson committed:
Switched back to three separate lists in place of single list

mike-w-wilson commented:
@lfrancioli , would you mind taking one last look as we went with the 3 array solution? Thanks!
